US771133: Support bulk cancellation of jobs

jonny-mccoubrey opened the pull request:
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=771133

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](http://sou-jenkins2.swinfra.net:8080/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=JobService&REPO_NAME=job-service&BRANCH_NAME=US771133) to go ahead and create the build...

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/JobService/job/JobService~job-service~US771133~CI

jonny-mccoubrey commented on job-service-db/src/main/resources/db/migration/functions/public/R__cancelJobs.sql:
```diff
@@ -0,0 +1,58 @@
+BEGIN
+    RETURN (
+        SELECT COUNT(cancel_job(in_partition_id, job_id))
+        FROM public.get_jobs(
```
Results from get_jobs are ordered as expected to avoid database deadlock:
https://github.com/JobService/job-service/blob/3ba44c579b27279b0ad836490f609b5b4a8d1473/job-service-db/src/main/resources/db/migration/functions/public/R__getJobs.sql#L212

jonny-mccoubrey commented:
Currently, when trying to cancel 1000 jobs, the endpoint is getting to 500 and stopping.

jonny-mccoubrey commented:
> Currently, when trying to cancel 1000 jobs, the endpoint is getting to 500 and stopping.

I was incrementing the `offset` parameter by the limit on each call of `cancel_jobs` from the `databaseHelper` however, because I was applying the 'NotFinished' filter, the `offset` wasn't using any jobs that had been cancelled already - so on every increment of 100 jobs I was missing 100 jobs hence only 500 of 1000 jobs were cancelled.

jonny-mccoubrey commented:
![image](https://github.com/JobService/job-service/assets/44871582/d92ad506-6779-44cf-906d-e7548a8fc61f)

Had to refactor the stored procedure so that get_jobs would return distinct job_ids when a job was being joined to `label` table.

Unfortunately, the function is now not as tidy.

jonny-mccoubrey commented:
@andyreidz good to know for review:

### Structure


**swagger.yaml**
Made contract changes to swagger file to generate boiler plate code for new endpoint

**JobsApi.java**
New endpoint added with parameters and responses (to be invoked by generated swagger code)

**JobsApiService.java**
Added new cancelJobs method

**JobsApiServiceImpl.java**
Implemented new cancelJobs method

**JobsCancel.java**
New method to prepare for call to database

**DatabaseHelper.java**
Added a new method to build SQL query and interact with the database in batched calls.
Separated out some code from getJobs that I am reusing for cancelJobs and could be used in the future for a deleteJobs endpoint.

**R__cancelJob.sql**
Modified this stored procedure to return a true if it completes successfully so that we can monitor the number of successful job cancellations.

**R__cancelJobs.sql**
Created a new stored procedure to cancel multiple jobs using the existing getJobs and cancelJob stored procedures.

andyreidz commented on job-service-db/src/main/resources/db/migration/functions/public/R__cancelJob.sql:
```diff
@@ -20,11 +20,13 @@
     in_job_id VARCHAR(48)
 )
-RETURNS VOID
+RETURNS BOOLEAN
```
Do we need this change since the function only ever returns true?

jonny-mccoubrey replied:
We don't - if there is an issue with cancelling a job, an exception will be raised.
Otherwise, the cancellation is a success and we increment the counter in cancel_jobs

andyreidz commented on job-service-db/src/main/resources/db/migration/functions/public/R__cancelJobs.sql:
```diff
@@ -0,0 +1,58 @@
+    counter integer;
+BEGIN
+    jobIds := ARRAY(SELECT DISTINCT job_id FROM public.get_jobs(in_partition_id, in_job_id_starts_with, in_status_type,
+        in_limit, in_offset, in_sort_field, in_sort_label, in_sort_ascending, in_labels, in_filter));
```
We need to order the items in the array by job id to avoid dead locks

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -137,10 +112,6 @@ public Job[] getJobs(final String partitionId, String jobIdStartsWith, String st
                     });
                 }
-            } finally {
-                if (array != null) {
```
You are leaking the array now

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -462,6 +433,41 @@ public void cancelJob(final String partitionId, String jobId) throws Exception {
+                final CallableStatement stmt = conn.prepareCall("{call cancel_jobs(?,?,?,?,?,?,?,?,?,?)}")
+        ) {
+            do {
+                final SortField sortField = JobSortField.CREATE_DATE;
```
I'm not sure buildGetJobsStatement is helping, it's introduced a memory leak and we don't need to pass sort field etc

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsCancel.java:
```diff
@@ -66,4 +70,45 @@ public static void cancelJob(final String partitionId, String jobId) throws Exce
+
+            LOG.debug("cancelJobs: Done");
+            return successfulCancellations;
+        } catch (Exception e) {
```
```suggestion
        } catch (final Exception e) {
```

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/generated/JobsApiService.java:
```diff
@@ -42,6 +42,8 @@ public abstract Response deleteJob(String partitionId,String jobId,String cAFCor
 
       public abstract Response cancelJob(String partitionId,String jobId,String cAFCorrelationId,SecurityContext securityContext)
               throws Exception;
+      public abstract Response cancelJobs(String partitionId, final String jobIdStartsWith, String label, final String filter,
```
final arguments

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/generated/JobsApiServiceImpl.java:
```diff
@@ -74,6 +73,15 @@ public Response cancelJob(final String partitionId, String jobId, String cAFCorr
     }
 
+    @Override
+    public Response cancelJobs(String partitionId, final String jobIdStartsWith, String label, final String filter,
```
final arguments

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -462,6 +433,41 @@ public void cancelJob(final String partitionId, String jobId) throws Exception {
+    public int cancelJobs(final String partitionId, String jobIdStartsWith, final List<String> labels, final String filter)
+            throws Exception {
+        int successfulCancellations = 0;
+        final int limit = 100;
```
Let's make this configurable through an ENV, just in case we need to tune it in production.

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -462,6 +433,41 @@ public void cancelJob(final String partitionId, String jobId) throws Exception {
+                        sortDirection, labels, filter, stmt, conn);
+
+                // Expect number of successful cancellations to be returned
+                stmt.registerOutParameter(1, Types.INTEGER);
```
Do we need this inside the loop?

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -591,4 +597,42 @@ private static void throwIfUnexpectedException(final SQLException se) throws Exc
+        stmt.setArray(9, array);
+        stmt.setString(10, filter);
+
+        if (array != null) {
```
I'm surprised this doesn't crash, because we are free'ing the array before we execute the statement.

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -462,6 +462,55 @@ public void cancelJob(final String partitionId, String jobId) throws Exception {
+            throws Exception {
+        int successfulCancellations = 0;
+
+        final String cancelBatchLimitEnvVar = System.getenv("CAF_CANCEL_JOBS_BATCH_LIMIT");
```
Have a look at AppConfig.java for examples on how to do this

andyreidz commented on job-service-db/src/main/resources/db/migration/functions/public/R__cancelJobs.sql:
```diff
@@ -0,0 +1,57 @@
+    counter integer;
+BEGIN
+    jobIds := ARRAY(SELECT DISTINCT job_id FROM public.get_jobs(in_partition_id, in_job_id_starts_with,
+        'NotFinished', in_limit, 0, 'create_date', null, false, in_labels, in_filter) ORDER BY job_id);
```
Sorry, should have said this earlier, let's use job_id for get_jobs sort field, sort descending like we do on ln44

andyreidz commented on job-service-db/src/main/resources/db/migration/functions/public/R__cancelJobs.sql:
```diff
@@ -0,0 +1,57 @@
+    FOREACH jobId IN ARRAY jobIds
+    LOOP
+        PERFORM cancel_job(in_partition_id, jobId);
+        counter := counter + 1;
```
In reality we don't need this counter as the return value will always be the length of the array.
If we encounter a failure an exception is raised and the transaction aborted.

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -462,6 +462,55 @@ public void cancelJob(final String partitionId, String jobId) throws Exception {
+    /**
+     * Cancels the specified jobs
+     */
+    public int cancelJobs(final String partitionId, String jobIdStartsWith, final List<String> labels, final String filter)
```
```suggestion
    public int cancelJobs(final String partitionId, final String jobIdStartsWith, final List<String> labels, final String filter)
```

jonny-mccoubrey replied:
If we make `jobIdStartsWith` final we can't set it to an empty string if it is null on L484.

andyreidz replied:
Fine, I see now that this pattern is used elsewhere in the existing code, I've made a suggestion below on how it could have been avoided.

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -462,6 +462,55 @@ public void cancelJob(final String partitionId, String jobId) throws Exception {
+                if (labels != null) {
+                    labelsArray = conn.createArrayOf("VARCHAR", labels.toArray());
+                } else {
+                    labelsArray = conn.createArrayOf("VARCHAR", new String[0]);
```
labelsArray.free ??

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsCancel.java:
```diff
@@ -66,4 +70,45 @@ public static void cancelJob(final String partitionId, String jobId) throws Exce
+
+            //  Get app config settings.
+            LOG.debug("cancelJobs: Reading database connection properties...");
+            AppConfig config = AppConfigProvider.getAppConfigProperties();
```
```suggestion
            final AppConfig config = AppConfigProvider.getAppConfigProperties();
```

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsCancel.java:
```diff
@@ -66,4 +70,45 @@ public static void cancelJob(final String partitionId, String jobId) throws Exce
+            AppConfig config = AppConfigProvider.getAppConfigProperties();
+
+            //  Get database helper instance.
+            DatabaseHelper databaseHelper = new DatabaseHelper(config);
```
```suggestion
            final DatabaseHelper databaseHelper = new DatabaseHelper(config);
```

andyreidz commented on job-service-db/src/main/resources/db/migration/functions/public/R__cancelJobs.sql:
```diff
@@ -0,0 +1,53 @@
+    jobIds varchar[];
+    jobId varchar;
+BEGIN
+    jobIds := ARRAY(SELECT DISTINCT job_id FROM public.get_jobs(in_partition_id, in_job_id_starts_with,
```
You still need
ORDER BY job_id

andyreidz commented on job-service/src/main/java/com/hpe/caf/services/job/api/DatabaseHelper.java:
```diff
@@ -462,6 +462,57 @@ public void cancelJob(final String partitionId, String jobId) throws Exception {
+                    jobIdStartsWith = "";
+                }
+                stmt.setString(1, partitionId);
+                stmt.setString(2, jobIdStartsWith);
```
```suggestion
                stmt.setString(2, jobIdStartsWith == null ? "" : jobIdStartsWith);
```

andyreidz replied:
No need to apply
